### PR TITLE
DPP-1468 Added Learner Builder Implementation

### PR DESCRIPTION
**Pull request by Pauljgraham**
…der Implementation and the Unit tests

### CODE REVIEW

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.csproj [hunk 1]
```diff
@@ -67,6 +67,7 @@
     <Compile Include="DependencyResolution\RefundsRegistry.cs" />
     <Compile Include="Domain\RefundGroup.cs" />
     <Compile Include="Dto\AccountLevyCredit.cs" />
+    <Compile Include="Dto\HistoricalPayment.cs" />
```
Comment (carlbm) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.csproj]: Can you move it to the Domain folder

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            IRequiredPaymentRepository requiredPaymentRepository,
+            IHistoricalPaymentsRepository historicalPaymentsRepository)
+        {
+            _requiredPaymentRepository = requiredPaymentRepository;
+            _historicalPaymentsRepository = historicalPaymentsRepository;
+        }
+
+        public List<LearnerData> CreateLearnersForProvider(long ukprn)
+        {
+            ResetLearnerResultsList();
```
Comment (carlbm) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerBuilder.cs]: Can you make this thread safe and make _learnerProcessParameters a local variable?

**Review by carlbm — CHANGES_REQUESTED** [review]

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.csproj (hunk 1 above)
Comment (Pauljgraham) [reply src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.csproj]: I'm confused on the difference between the domain folder, and the dto folder. Happy to do this but would like to discuss the purpose of each of these folders.

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/HistoricalPayment.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
+﻿using SFA.DAS.Payments.DCFS.Domain;
+using SFA.DAS.ProviderPayments.Calc.Refunds.Infrastructure.Entities;
+using SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.ProviderPayments.Calc.Refunds.Domain
+{
+    public class HistoricalPayment : PaymentEntity
```
Comment (scottywakefield) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/HistoricalPayment.cs:7]: Is `PaymentEntity` an entity or a domain model? if an entity then this should be changed as we don't want a domain model inheriting from a data entity (seems wrong to me anyway). if its a domain model then it should be renamed.

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/RefundGroup.cs [hunk 4]
```diff
@@ -13,7 +14,7 @@ public RefundGroup(RequiredPaymentEntity entity)
             ApprenticeshipContractType = entity.ApprenticeshipContractType;
         }
 
-        public RefundGroup(HistoricalPaymentEntity entity)
+        public RefundGroup(HistoricalPayment entity)
```
Comment (scottywakefield) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/RefundGroup.cs:17]: check your renames, might be a few other var names that need updating like this one, from `entity` to something `historicalPayment` or similar.

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+        public LearnerBuilder(
+            IRequiredPaymentRepository requiredPaymentRepository,
+            IHistoricalPaymentsRepository historicalPaymentsRepository)
+        {
+            _requiredPaymentRepository = requiredPaymentRepository;
+            _historicalPaymentsRepository = historicalPaymentsRepository;
+        }
+
+        public List<LearnerData> CreateLearnersForProvider(long ukprn)
```
Comment (scottywakefield) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerBuilder.cs:24]: A suggestion - if LearnerData has a domain model for one data type (`HistoricalPayment`) but a data entity for the other (`RequiredPaymentEntity`) then it is inconsistent and should use all domain models.

File: src/Shared/SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure/Data/Entities/HistoricalPaymentEntity.cs [hunk 6]
```diff
@@ -1,25 +0,0 @@
-﻿using System.ComponentModel.DataAnnotations;
-
-namespace SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure.Data.Entities
-{
-    public class HistoricalPaymentEntity : PaymentEntity
```
Comment (scottywakefield) [inline src/Shared/SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure/Data/Entities/HistoricalPaymentEntity.cs:5]: What happened to `HistoricalPaymentsEntity`? was this a duplicate?

**Review by scottywakefield — CHANGES_REQUESTED** [review]
A few minor comments on moving from an entity to domain model.

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/HistoricalPayment.cs (hunk 3 above)
Comment (Pauljgraham) [reply src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/HistoricalPayment.cs:7]: This code is taken from the 'Make a payment' branch (the one we didn't get time to review.) It's far from ideal that we have a domain object having a dependency on an Entity (however we have done this with the refund domain object).  Please discuss with Carl and see his view

File: src/Shared/SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure/Data/Entities/HistoricalPaymentEntity.cs (hunk 6 above)
Comment (Pauljgraham) [reply src/Shared/SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure/Data/Entities/HistoricalPaymentEntity.cs:5]: yes, you added it in the local refunds repo and Carl added it in the shared repo

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/RefundGroup.cs (hunk 4 above)
Comment (Pauljgraham) [reply src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/RefundGroup.cs:17]: Good spot, missed that

Comment (Pauljgraham) [reply src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/RefundGroup.cs:17]: Actually, I just looked again. The parameter Type in the RefundGroup needs to use the domain object and not the entity from the refunds repo

Comment (Pauljgraham) [reply src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/RefundGroup.cs:17]: It was resharper which renamed these refs

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerBuilder.cs (hunk 5 above)
Comment (Pauljgraham) [reply src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerBuilder.cs:24]: Agreed, but again lets talk to Carl. This again is to do with inconsistent styles. And allowing the use of entities in the domain

**Review by carlbm — APPROVED** [review]

**Review by scottywakefield — APPROVED** [review]

**Comment by scottywakefield** [comment]
Note: we need to revisit segregation of entities and domain models.

**Pull request merged**